pandas/core/construction.py

```diff
@@ -275,28 +276,28 @@ def array(
     ):
         dtype = data.dtype
 
-    data = extract_array(data, extract_numpy=True)
+    _data = extract_array(data, extract_numpy=True)
```

Comment:
Why is this needed? What is returned here should still be typed the same as the original `data`

Reply:
> What is returned here should still be typed the same as the original data

IIUC the return type is a numpy array. And therefore data should be narrowed to just np.ndarray. This is not the same as the original data, which is typed as Union[Sequence[object], AnyArrayLike]

The crux of issue is np.ndarray resolves to Any. 

data is typed Union[Sequence[object], AnyArrayLike] which is equivalent to Union[Sequence[object], Any, Index, Series, ExtensionArray].

np.ndarray resolving to Any prevents mypy being able to type narrow from extract_array for a few reasons.

1. if we overload extract_array we get `error: Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader`.

```
@overload
def extract_array(
    obj: Union[Sequence[object], AnyArrayLike], extract_numpy: bool = False
) -> ArrayLike:
    ...


@overload
def extract_array(obj: Any, extract_numpy: bool = False) -> Any:
    ...


def extract_array(obj: Any, extract_numpy: bool = False) -> Union[Any, ArrayLike]:
```

AnyArrayLike includes np.ndarray which resolves to Any and therefore the second signature is unreachable. 


2. even if np.ndarray did not resolve to Any, without Literal (py 3.8) we can't overload extract_numpy to return a numpy array i.e to implicitly `cast` to the return type of extract_array, i.e. from extract_array(Union[Sequence[object], AnyArrayLike], extract_numpy=True) -> np.ndarray ( requires additional overloads to above)

3. without the overloads, we have just a union return type from extract_array. we can't cast from Union[Sequence[object], AnyArrayLike] to Union[Any, ArrayLike] without a mypy error `pandas\core\construction.py:279: error: Redundant cast to "Any"` (unless we remove warn_redundant_casts = True from setup.cfg) 

so in summary, _data is a separate type to data that resolves for Any since it is a numpy array following the extract_array call and we revert to dynamic typing for the rest of the method.